AndroidTV Config Flow (#20787)
home-assistant/home-assistant.github.io: source/_integrations/androidtv.markdown

  - From the main (Launcher) screen, select Settings.
  - Select My Fire TV > About > Network.

{% include integrations/config_flow.md %}

## Integration Options

> It is possible to change some behaviors through the integration options. These can be changed at **AndroidTV** -> **Configure** on the Integrations page.

Reviewer: We have helpers for these as well, see the `cast` documentation for an example

  - From the main (Launcher) screen, select Settings.
  - Select My Fire TV > About > Network.

{% include integrations/config_flow.md %}

{% include integrations/option_flow.md %}
{% configuration_basic %}
Configure Applications List:

>   description: Here you can define a list of applications where the keys are app IDs and the values are app names that will be displayed in the UI. If a name is not provided, the app will never be shown in the sources list. ([These app names](https://github.com/JeffLIrion/python-androidtv/blob/748d6b71cad611c624ef7526d9928431167531a3/androidtv/constants.py#L290-L308) are configured in the backend package and do not need to be included in your configuration).

Reviewer 2: > If a name is not provided, the app will never be shown in the sources list.

That is only true if `exclude_unnamed_apps` (or its successor) is true. And I believe it is false by default.

  - From the main (Launcher) screen, select Settings.
  - Select My Fire TV > About > Network.

{% include integrations/config_flow.md %}

{% include integrations/option_flow.md %}
{% configuration_basic %}
Configure Applications List:

>   description: Here you can define a list of applications where the keys are app IDs and the values are app names that will be displayed in the UI. If a name is not provided and the option `Exclude apps with unknown name` is enabled, the app will never be shown in the sources list. ([These app names](https://github.com/JeffLIrion/python-androidtv/blob/748d6b71cad611c624ef7526d9928431167531a3/androidtv/constants.py#L290-L308) are configured in the backend package and do not need to be included in your configuration).

Reviewer 2: ```suggestion
  description: Here you can define a list of applications where the keys are app IDs and the values are app names that will be displayed in the UI. If a name is not provided and the option `Exclude apps with unknown name` is enabled, the app will never be shown in the sources list. ([These app names](https://github.com/JeffLIrion/python-androidtv/blob/v0.0.60/androidtv/constants.py#L396-L482) are configured in the backend package and do not need to be included in your configuration).
```

Reviewer: Honestly, we should not be linking to source code from end-user documentation AT ALL.

Author: There are other points where should code is referenced. should we remove everywhere? What you suggest, to generally link to the library repository?

Reviewer: > What you suggest, to generally link to the library repository?

I suggest that the needed configuration options are documented in the Home Assistant documentation.

  description: Here you can configure a list of rules where the rule key is the app IDs and whose values are lists of state detection rules. As example a valid value for a detection rule is `["standby", {"playing":{"media_session_state":4}}, {"paused":{"media_session_state":3, "wake_lock_size":4}}]`. Note that rule values must be always inside square bracket (`[...]`). See the section [Custom State Detection](#custom-state-detection) for more info.
{% endconfiguration_basic %}

## ADB Setup

This integration works by sending ADB commands to your Android TV / Fire TV device. There are two ways to accomplish this.

<div class='note'>

When connecting to your device for the first time, a dialog will appear on your Android TV / Fire TV asking you to approve the connection. Check the box that says "always allow connections from this device" and hit OK.

</div>

### 1. Python ADB Implementation

The default approach is to connect to your device using the `adb-shell` Python package. As of Home Assistant 0.101, if a key is needed for authentication and it is not provided by the `ADB Key` setup option, then Home Assistant will generate a key for you.

<div class='note'>

> To be able to provide `ADB Key` on integration setup, you need to enable [advanced mode](https://www.home-assistant.io/blog/2019/07/17/release-96/#advanced-mode).

Reviewer: ```suggestion
To be able to provide `ADB Key` on integration setup, you need to enable [advanced mode](/blog/2019/07/17/release-96/#advanced-mode).
```

</div>

Prior to Home Assistant 0.101, this approach did not work well for newer devices. Efforts have been made to resolve these issues, but if you experience problems then you should use the ADB server option.

### 2. ADB Server

The second option is to use an ADB server to connect to your Android TV and Fire TV devices.

<div class='note'>

> To configure ADB server on integration setup, you need to enable [advanced mode](https://www.home-assistant.io/blog/2019/07/17/release-96/#advanced-mode).

Reviewer: ```suggestion
To configure ADB server on integration setup, you need to enable [advanced mode](/blog/2019/07/17/release-96/#advanced-mode).
```

### 1. Python ADB Implementation

The default approach is to connect to your device using the `adb-shell` Python package. As of Home Assistant 0.101, if a key is needed for authentication and it is not provided by the `ADB Key` setup option, then Home Assistant will generate a key for you.

> <div class='note'>
> To be able to provide `ADB Key` on integration setup, you need to enable [advanced mode](https://www.home-assistant.io/blog/2019/07/17/release-96/#advanced-mode).
> </div>

Reviewer: Adjustment to the link and an empty line before after language switching to help the Markdown parse to switch HTML/Markdown context.

```suggestion
<div class='note'>

To be able to provide `ADB Key` on integration setup, you need to enable [advanced mode](/blog/2019/07/17/release-96/#advanced-mode).

</div>
```

Prior to Home Assistant 0.101, this approach did not work well for newer devices. Efforts have been made to resolve these issues, but if you experience problems then you should use the ADB server option.

### 2. ADB Server

The second option is to use an ADB server to connect to your Android TV and Fire TV devices.

> <div class='note'>
> To configure ADB server on integration setup, you need to enable [advanced mode](https://www.home-assistant.io/blog/2019/07/17/release-96/#advanced-mode).
> </div>

Reviewer: ```suggestion
<div class='note'>

To configure ADB server on integration setup, you need to enable [advanced mode](/blog/2019/07/17/release-96/#advanced-mode).

</div>
```

### 1. Python ADB Implementation

> The default approach is to connect to your device using the `adb-shell` Python package. As of Home Assistant 0.101, if a key is needed for authentication and it is not provided by the `ADB Key` setup option, then Home Assistant will generate a key for you.

Reviewer 2: Is providing an ADB key a hidden/advanced configuration option? After all, it's easier and simpler to just let HA generate a key.

Author: Yes is advanced option, and is specified in the next chapter `Python ADB Implementation`
